securedrop/journalist_app/sessions.py

```diff
@@ -0,0 +1,283 @@
+                session["renew_count"] -= 1
+                expires += self.lifetime
+                session.modified = True
+        conditional_cookie_kwargs = {}
+        httponly = self.get_cookie_httponly(app)
+        secure = self.get_cookie_secure(app)
+        if self.has_same_site_capability:
+            conditional_cookie_kwargs["samesite"] = self.get_cookie_samesite(app)
+        val = self.serializer.dumps(dict(session))
+        if session.to_regenerate:
+            self.redis.delete(self.key_prefix + session.sid)
+            session.sid = self._generate_sid()
+            session.token = self._get_signer(app).dumps(session.sid)  # type: ignore
+        if session.new or session.to_regenerate:
+            self.redis.setex(name=self.key_prefix + session.sid, value=val, time=expires)
+        elif session.modified:
+            # To prevent race conditions where session is delete by an admin in the middle of a req
+            # accept to save the session object if and only if alrady exists using the xx flag
+            self.redis.set(name=self.key_prefix + session.sid, value=val, ex=expires, xx=True)
+        if not session.is_api and (session.new or session.to_regenerate):
+            response.headers.add("Vary", "Cookie")
+            response.set_cookie(
+                app.session_cookie_name,
+                session.token,
+                httponly=httponly,
+                domain=domain,
+                path=path,
+                secure=secure,
+                **conditional_cookie_kwargs  # type: ignore
+            )
+
+
+class Session:
+    def __init__(self, app: Flask) -> None:
+        self.app = app
+        if app is not None:
+            self.init_app(app)
+
+    def init_app(self, app: Flask) -> "None":
+        """This is used to set up session for your app object.
+        :param app: the Flask app object with proper configuration.
+        """
+        app.session_interface = self._get_interface(app)  # type: ignore
+
+    def _get_interface(self, app: Flask) -> SessionInterface:
+        config = app.config.copy()
+        config.setdefault("SESSION_REDIS", Redis())
+        config.setdefault("SESSION_LIFETIME", 2 * 60 * 60)
+        config.setdefault("SESSION_RENEW_COUNT", 5)
+        config.setdefault("SESSION_SIGNER_SALT", "session")
+        config.setdefault("SESSION_KEY_PREFIX", "session:")
+        config.setdefault("SESSION_HEADER_NAME", "authorization")
+
+        session_interface = SessionInterface(
+            config["SESSION_LIFETIME"],
+            config["SESSION_RENEW_COUNT"],
+            config["SESSION_REDIS"],
+            config["SESSION_KEY_PREFIX"],
+            config["SESSION_SIGNER_SALT"],
+            config["SESSION_HEADER_NAME"],
+        )
+
+        return session_interface
+
+
+def logout_user(uid: int) -> None:
+    redis = Redis()
+    for key in redis.keys(app.config["SESSION_KEY_PREFIX"] + "*") + redis.keys(
```

Comment:
This config key may not be set on existing instances that were upgraded. Calling this function causes a 500 error with the following example stack trace: 
```
[Fri Sep 23 18:35:04.907262 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948] E
RROR:journalist_app:Exception on /account/new-password [POST]                                          
[Fri Sep 23 18:35:04.907331 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948] T
raceback (most recent call last):                                                                      
[Fri Sep 23 18:35:04.907354 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948]  
 File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 2073, in wsgi_app
[Fri Sep 23 18:35:04.907370 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948]  
   response = self.full_dispatch_request()                                                             
[Fri Sep 23 18:35:04.907384 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948]  
 File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 1518, in full_dis
patch_request                                                                                          
[Fri Sep 23 18:35:04.907414 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948]  
   rv = self.handle_user_exception(e)                                                                  
[Fri Sep 23 18:35:04.907428 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948]   File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 1516, in full_dispatch_request            
[Fri Sep 23 18:35:04.907442 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948]     rv = self.dispatch_request()                    
[Fri Sep 23 18:35:04.907454 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948]   File "/opt/venvs/securedrop-app-code/lib/python3.8/site-packages/flask/app.py", line 1502, in dispatch_request                 
[Fri Sep 23 18:35:04.907467 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948]     return self.ensure_sync(self.view_functions[rule.endpoint])(**req.view_args)                        
[Fri Sep 23 18:35:04.907478 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948]   File "/var/www/securedrop/journalist_app/account.py", line 46, in new_password                        
[Fri Sep 23 18:35:04.907486 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948]     logout_user(user.id)  
[Fri Sep 23 18:35:04.907498 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948]
 File "/var/www/securedrop/journalist_app/sessions.py", line 262, in logout_user                       
[Fri Sep 23 18:35:04.907511 2022] [wsgi:error] [pid 687:tid 118508390467328] [remote 127.0.0.1:60948] KeyError: 'SESSION_KEY_PREFIX'  
```

Reply:
As `logout_user` was the only function outside the class it would thus not fallback on the default values. Fix is to move it in the same context as the other functions that require `self.key_prefix` and `self.redis`. It is also cleaner since we do not reinstantiate the redis object too. https://github.com/freedomofpress/securedrop/pull/6403/commits/64d40726c8f611b137dcfbdb76caf9e6f955bfa4